nativelink-store/src/redis_store.rs

```diff
@@ -365,6 +383,58 @@ impl StoreDriver for RedisStore {
             .await
     }
 
+    async fn list(
+        self: Pin<&Self>,
+        range: (Bound<StoreKey<'_>>, Bound<StoreKey<'_>>),
+        handler: &mut (dyn for<'a> FnMut(&'a StoreKey) -> bool + Send + Sync + '_),
+    ) -> Result<u64, Error> {
+        let range = (
+            range.0.map(StoreKey::into_owned),
+            range.1.map(StoreKey::into_owned),
+        );
+        let pattern = match range.0 {
+            Bound::Included(ref start) | Bound::Excluded(ref start) => match range.1 {
+                Bound::Included(ref end) | Bound::Excluded(ref end) => {
+                    let start = start.as_str();
+                    let end = end.as_str();
+                    let max_length = start.len().min(end.len());
+                    let length = start
+                        .chars()
+                        .zip(end.chars())
+                        .position(|(a, b)| a != b)
+                        .unwrap_or(max_length);
+                    format!("{}{}*", self.key_prefix, &start[..length])
+                }
+                Bound::Unbounded => format!("{}*", self.key_prefix),
+            },
+            Bound::Unbounded => format!("{}*", self.key_prefix),
+        };
+        let client = self.client_pool.next();
+        let mut scan_stream = client.scan(pattern, Some(self.scan_count), None);
+        let mut iterations = 0;
+        'outer: while let Some(mut page) = scan_stream.try_next().await? {
```

Comment:
The unexpected protocol error from `fred` may be suppressed but then there is no path for errors to propagate.
```suggestion
        'outer: while let Some(mut page) = scan_stream.try_next().await.unwrap_or(None) {
```

Reply:
I confirmed that backporting this implementation to `v0.5.3` works as expected. That release's `Cargo.lock` is using `fred 9.1.2`, which precedes the changes to scan in `fred 10.0.0`. I will try to isolate the issue and make a bug report.

Reply:
Oh, great. I'm sure the `fred` devs will appreciate it. As you know, `0.5.x` in NativeLink is not advised. Maybe we can downgrade the `fred` version, depending on why we upgraded it. 

Thank you.

Reply:
It happens.